src/Libraries/Microsoft.Extensions.AI/CachingHelpers.cs

```diff
@@ -36,19 +36,41 @@ public static string GetCacheKey(ReadOnlySpan<object?> values, JsonSerializerOpt
         // invalidating any existing cache entries that may exist in whatever IDistributedCache was in use.
 
 #if NET
-        IncrementalHashStream? stream = IncrementalHashStream.ThreadStaticInstance ?? new();
-        IncrementalHashStream.ThreadStaticInstance = null;
+        IncrementalHashStream? stream = IncrementalHashStream.ThreadStaticInstance;
+        if (stream is not null)
+        {
+            // We need to ensure that the value in ThreadStaticInstance is always ready to use.
+            // If we start using an instance, write to it, and then fail, we will have left it
+            // in an inconsistent state. So, when renting it, we null it out, and we only put
+            // it back upon successful completion after resetting it.
+            IncrementalHashStream.ThreadStaticInstance = null;
```

Comment:
Why not just dispose it in a try/finally? Is it expected that this could be called recursively?

Reply:
It shouldn't be disposed if it's going to be cached. So it's disposed in a catch rather than a finally.